client/tokenized-payment-request/payment-request.js

```diff
@@ -0,0 +1,406 @@
+/* global jQuery, wcpayPaymentRequestParams */
+/**
+ * External dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { doAction, addAction, removeAction } from '@wordpress/hooks';
+
+/**
+ * Internal dependencies
+ */
+import {
+	setPaymentRequestBranding,
+	trackPaymentRequestButtonClick,
+	trackPaymentRequestButtonLoad,
+} from './tracking';
+import {
+	transformStripePaymentMethodForStoreApi,
+	transformStripeShippingAddressForStoreApi,
+} from './transformers/stripe-to-wc';
+import {
+	transformCartDataForDisplayItems,
+	transformCartDataForShippingOptions,
+} from './transformers/wc-to-stripe';
+import paymentRequestButtonUi from './button-ui';
+import {
+	getPaymentRequest,
+	displayLoginConfirmationDialog,
+	getPaymentRequestData,
+	waitForAction,
+} from './frontend-utils';
+import PaymentRequestCartApi from './cart-api';
+import debounce from './debounce';
+
+const noop = () => null;
+
+/**
+ * Class to handle Stripe payment forms.
+ */
+export default class WcpayPaymentRequest {
+	/**
+	 * Whether the payment was aborted by the customer.
+	 */
+	paymentAborted = false;
+
+	/**
+	 * Whether global listeners have been added.
+	 */
+	areListenersInitialized = false;
+
+	/**
+	 * The cart data represented if the product were to be added to the cart (or, on cart/checkout pages, the cart data itself).
+	 * This is useful on product pages to understand if shipping is needed.
+	 */
+	cachedCartData = undefined;
+
+	/**
+	 * API to interface with the cart.
+	 *
+	 * @type {PaymentRequestCartApi}
+	 */
+	paymentRequestCartApi = undefined;
+
+	/**
+	 * WCPayAPI instance.
+	 *
+	 * @type {WCPayAPI}
+	 */
+	wcpayApi = undefined;
+
+	/**
+	 * On page load for product pages, we might get some data from the backend (which might get overwritten later).
+	 */
+	initialProductData = undefined;
+
+	constructor( { wcpayApi, paymentRequestCartApi, productData } ) {
+		this.wcpayApi = wcpayApi;
+		this.paymentRequestCartApi = paymentRequestCartApi;
+		this.initialProductData = productData;
+	}
+
+	/**
+	 * Starts the payment request
+	 */
+	async startPaymentRequest() {
+		// reference to this class' instance, to be used inside callbacks to avoid `this` misunderstandings.
+		const _self = this;
+		// TODO: is this creating multiple handlers to events on different `paymentRequest` objects?
+		const paymentRequest = getPaymentRequest( {
+			stripe: this.wcpayApi.getStripe(),
+			cartData: this.cachedCartData,
+			productData: this.initialProductData,
+		} );
+
+		// Check the availability of the Payment Request API first.
+		const paymentPermissionResult = await paymentRequest.canMakePayment();
+		if ( ! paymentPermissionResult ) {
+			doAction( 'wcpay.payment-request.availability', {
+				paymentRequestType: null,
+			} );
+			return;
+		}
+
+		// TODO: Don't display custom button when paymentRequestType is `apple_pay` or `google_pay`.
```

Comment:
I am not sure, to be honest 😬 I copied the `TODO` and the logic from https://github.com/Automattic/woocommerce-payments/blob/654097724c429f2ba812b94e7094c2c0f3b1df33/client/payment-request/index.js#L243-L243

Reply:
This TODO looks like it was from a really long time ago of yore. It was added in #1811, where it looks there was a third option for the `payment_request_type` where we just presented a custom button.  Do we still actually support a custom button like this in our current UI? I don't believe I've ever seen it before, but that's not necessarily exceptional. 

[Stripe suggests](https://docs.stripe.com/js/payment_request/can_make_payment) that the three payment method request types returned here should either be Apple, Google, or Link. AFAIA, we do not support Link in the payment request button...so do we even still need this third option at all? 🤔

Reply:
> Do we still actually support a custom button like this in our current UI? 

No, we don't support that

> AFAIA, we do not support Link in the payment request button...so do we even still need this third option at all? 🤔

To confirm - are you recommending simplifying the logic to just checking for either `applePay` or `googlePay`? Something like this?

```js
const buttonBranding = paymentPermissionResult.applePay ? 'apple_pay' : 'google_pay';
```

Reply:
> To confirm - are you recommending simplifying the logic to just checking for either applePay or googlePay? Something like this?

Yeah, basically. Again, I'm not sure if there is any historical context for the third `payment_request_api` being there from the beginning, if it's still a required fallback, or what might break if we were to remove it. That being said, I'd also rather not leave it in, just because we're unsure what it's supposed to do. 😅 

Here are a couple considerations things that I believe are also worth mentioning. First of all, in the current implementation, if the third `payment_request_api` option is somehow chosen, it will be ignored by our tracking in any case, since it isn't an included value in the array over [here](https://github.com/Automattic/woocommerce-payments/blob/93bdadfe0fffb925d7b2b82b89996243c4d7a8fa/client/tokenized-payment-request/tracking.js#L12-L13).

Secondly, it's likely that we will attempt to support Link--and maybe more wallet payment methods--with the ECE in the future, which this implementation should be capable of supporting with hopefully as minimal adaptation as possible. It would probably be helpful to store all of these string key constants (e.g. `apple_pay`, `google_pay`, `link`, etc.) in one place and refer back to that one constant store in all places where these strings are required, so that if we need to add more strings to the list of supported express payment methods, we could just update a single location without having to hunt around the code and face these similar concerns about which strings are valid. 

Any thoughts about any of the above? 🤔

Reply:
While looking at other opportunities to improve this, I noticed that there is this other instance of `payment_request_api`: https://github.com/Automattic/woocommerce-payments/blob/develop/client/payment-request/utils/utils.js#L146-L146

And the `.replace(...)` is used to replace the payment method in this string: https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payments-payment-request-button-handler.php#L1476

So again, it seems to be a sub-optimal experience for this `payment_request_api` value.

I'd be in favor of getting rid of any references to this `payment_request_api` string, and just assuming either `apple_pay` or `google_pay`.
As you mentioned, [the Stripe docs](https://docs.stripe.com/js/payment_request/can_make_payment) state that the return object could either have `applePay`, `googlePay`, or `link`. But you can't have more than one of those attributes "truth".

Given that the ECE could have [more than one payment method enabled at the same time](https://docs.stripe.com/js/elements_object/create_express_checkout_element#express_checkout_element_create-options-paymentMethods), and that a new implementation is necessary, I'd be in favor of getting rid of this `payment_request_api` reference 🤷

Reply:
I feel relatively confident that this `payment_request_api` constant is no longer relevant and can probably be removed from our implementation. However, I'm equally happy to investigate this more comprehensively in a separate issue, as it doesn't really appear to affect current functionality if it's present or not. Can we just perhaps update the TODO to make it more meaningful and relevant to the current confusion--perhaps just to investigate the purpose of the `payment_request_api` constant--and then we can address all of these TODOs separately in a new issue?